lib/puppet/configurer.rb

```diff
@@ -217,26 +217,22 @@ def run(options = {})
 
         # Skip failover logic if the server_list setting is empty
         if Puppet.settings[:server_list].nil? || Puppet.settings[:server_list].empty?
-          do_failover = false;
+          do_failover = false
         else
           do_failover = true
         end
         # When we are passed a catalog, that means we're in apply
         # mode. We shouldn't try to do any failover in that case.
         if options[:catalog].nil? && do_failover
-          found = find_functional_server()
-          server = found[:server]
-          if server.nil?
-            Puppet.warning _("Could not select a functional puppet master")
-            server = [nil, nil]
-          end
-          Puppet.override(:server => server[0], :serverport => server[1]) do
-            if !server.first.nil?
-              Puppet.debug "Selected master: #{server[0]}:#{server[1]}"
-              report.master_used = "#{server[0]}:#{server[1]}"
+          server, port = find_functional_server
+          Puppet.override(server: server, serverport: port) do
```

Comment:
If there are no functional servers, this PR preserves the behavior of pushing a `nil` server value onto the context. Later when we try to resolve the server, we see that the `:server` key exists, and the context system never calls the [block to resolve the default server](https://github.com/puppetlabs/puppet/blob/2b83deb189a4104ecea2fdcb9151c430afd72d9b/lib/puppet/util/connection.rb#L28-L35). That is the cause of https://tickets.puppetlabs.com/browse/PUP-8037. I didn't want to fix that in this PR, because there are several issues with `server_list`.